### PR TITLE
Fix several issues regarding Typescript 2.0 compiler flags and new typings version.

### PR DESCRIPTION
**Pull request by remolueoend**

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -100,6 +113,7 @@ declare class Server implements NodeJS.EventEmitter {
 
     addListener(event: string, listener: Function): this;
     on(event: string, listener: Function): this;
+    on(event: 'error', listener: ErrorCallback): this;
```
Comment (freeman) [inline index.d.ts:120]: adding typed support for the other events would be great

Comment (remolueoend) [reply index.d.ts:120]: thanks for the feedback, I'm on it 👍🏻

**Comment by SomaticIT** [comment]
Thanks @remolueoend @freeman for your work.

@remolueoend Do you want me to wait for the others events' typings before merging this PR ?

**Comment by remolueoend** [comment]
Sorry for my late response. I added all definitions of event types I could find in the original source code. Some remarks:
* `target` parameter in `ErrorCallback` might be missing (see: https://github.com/nodejitsu/node-http-proxy/blob/d0f1dfeb8277d46a057017cd888b50e85f6725d6/lib/http-proxy/index.js#L68) or might be a string or a parsed URL-object, because the string based URL provided via options is not parsed until `createRightProxy` (https://github.com/nodejitsu/node-http-proxy/blob/d0f1dfeb8277d46a057017cd888b50e85f6725d6/lib/http-proxy/index.js#L64)

* Both client-requests and target-responses (`proxyRes`) in perspective of the proxy-server are of type `http.IncomingMessage`.

* Type of `socket` in websocket related events strongly depends on the use-case. Based on the examples of the original README of node-http-proxy, `socket` should be a `net.Socket` in most cases:
```javascript
//
// Listen to the `upgrade` event and proxy the
// WebSocket requests as well.
//
proxyServer.on('upgrade', function (req, socket, head) {
  proxy.ws(req, socket, head);
});
```

Best solution might be using generic types, eg:
```typescript
proxy.ws<TSocket>(req: http.IncomingMessage, socket: TSocket, head: any)
```

**Comment by SomaticIT** [comment]
Indeed we can use a generic type for this case.

**Comment by SomaticIT** [comment]
I merge this first batch. We could add generic version of `ws` method in a new PR.
Thanks for your work !

**Pull request merged**